Module 3: CRUD and data models (#4)
ReCoded-Org/curriculum-backend-readings: module3-crud-and-data-models/r3-model-view-controller/README.md

# Model View Controller (MVC)

## Introduction to MVC projects

> When beginning a new project, there are several different methods that can be used to set up the overall file structure and flow. One of the most commonly used architectural patterns is called MVC. This is an acronym for `Model`, `View`, `Controller`.

Reviewer: this acronym stands for "Model, View, Controller"

Reviewer: and again not using `code` formatting

> This pattern is favored because of its alignment with the computer science design principle, Separation of Concerns. By dividing up responsibilities within our file structure, we can encapsulate information to be referred to via abstraction and maintain cleaner codebases.

Reviewer: bold, or include a link, to something about separation of concerns

Reviewer: "By dividing up responsibilities within our file structure, we can encapsulate information to be referred to via abstraction and maintain cleaner codebases." 

also this needs to be made a much better explanation.

```
app-name-here
 |- Controllers
 |- db
 |- Models
 |- node_modules
.gitignore
index.js
package.json
package-lock.json
```

> ### Models

Reviewer: above this line or above the file structure you should give a brief one sentence summary of what each part of MVC is, then in each subsection go into it in more detail

and again explaining that MVC is not a super rigid thing

> In the models folder, the files will contain our schema definition and expected fields for the model.

Reviewer: it's not explained anywhere what a "model" generally is

The naming convention for the model files is: `name-here-model.js`.

Here’s an example model using JavaScript and mongoose.

```js
// post-model.js

const mongoose = require("mongoose");

const post = new mongoose.Schema({
  title: {
    type: String,
    required: true,
  },
  body: {
    type: String,
    required: true,
  },
  createdAt: {
    type: Date(),
  },
  author: {
    type: String,
    required: true,
  },
});

module.exports = mongoose.model("Post", post);
```

### Controllers

> Controllers are the layer between the Model and the View. the views can use the controllers to `add`, `read`, `delete`, ...etc data.

Reviewer: i think this explanation needs to be much better of what a controller is.

Reviewer 2: it is worth mentioning that the controller can be used to sanitize and validate the requests before sending them to the model data validators.

Ref: https://stackoverflow.com/a/5305911/5070971

The naming convention for the model files is: `name-here-controller.js`.

Here is an example:

```js
// post-controller.js

const express = require('express');
const router = express.Router();
const Post = require('../models/post-model.js')

router.get('/posts', (req, res) => {
 Post.find({})
  .then((posts) => {
    res.json(posts);
  })
  .catch(console.error)
})
});
```

### Views

> In a back-end application, views are usually not implemented and rather we create a front-end app using maybe `React` to call our `api` end-points to manipulate the data in the back-end.

Reviewer: this explanation also does not really teach someone what MVC is

Reviewer: like, what is a view?

# Model View Controller (MVC)

## Introduction to MVC projects

> When beginning a new project (especially, when using express), there are several different methods that can be used to set up the overall file structure and flow. One of the most commonly used architectural patterns is called MVC. This acronym stands for "Model, View, Controller".

Reviewer: ```suggestion
When beginning a new project, especially when using Express, there are several different methods that can be used to set up the overall file structure and flow. One of the most commonly used architectural patterns is called MVC. This acronym stands for "Model, View, Controller".
```

> This pattern is favored because of its alignment with the computer science design principle, [**separation of concerns**](https://en.wikipedia.org/wiki/Separation_of_concerns). By dividing up responsibilities within our file structure, for example, we can have our db connection work in one file and api routes in another file, ...etc.

Reviewer: ```suggestion
This pattern is favored because of its alignment with the computer science design principle, [**separation of concerns**](https://en.wikipedia.org/wiki/Separation_of_concerns). By dividing up responsibilities within our file structure, for example, we can have our database connection work in one file and API routes in another file, and so on. This makes our code better organized, which ultimately helps ease the developer workflow and prevent bugs.
```

```
app-name-here
 |- Controllers
 |- db
 |- Models
 |- node_modules
.gitignore
index.js
package.json
package-lock.json
```

### Components Of MVC

**Model:** This consists of the structure of our data and handle the database.

**View:** The part of our application which is shown to the user.

**Controller:** Controller has all the logic to control and respond to the action, the user performs using the views.

### Models

> In this folder, you can write the functionality & logics related to the Database (if you aren't using ORM) like insert, fetch, update, delete queries. It takes the query request from the controller & sends the response back to the controller.

Reviewer: database randomly capitalized

## Introduction to MVC projects

> When beginning a new project (especially, when using express), there are several different methods that can be used to set up the overall file structure and flow. One of the most commonly used architectural patterns is called MVC. this acronym stands for "Model, View, Controller".

Reviewer: ```suggestion
When beginning a new project (especially when using express), there are several different methods that can be used to set up the overall file structure and flow. One of the most commonly used architectural patterns is called MVC. this acronym stands for "Model, View, Controller".
```

> When beginning a new project (especially, when using express), there are several different methods that can be used to set up the overall file structure and flow. One of the most commonly used architectural patterns is called MVC. this acronym stands for "Model, View, Controller".

Reviewer: ```suggestion
When beginning a new project (especially, when using express), there are several different methods that can be used to set up the overall file structure and flow. One of the most commonly used architectural patterns is called MVC. This acronym stands for "Model, View, Controller".
```

> This pattern is favored because of its alignment with the computer science design principle, [**separation of concerns**](https://en.wikipedia.org/wiki/Separation_of_concerns). By dividing up responsibilities within our file structure, for example, we can have our db connection work in one file and api routes in another file, ...etc.

Reviewer: ```suggestion
This pattern is commonly found because of its alignment with the computer science design principle, [**separation of concerns**](https://en.wikipedia.org/wiki/Separation_of_concerns). By dividing up responsibilities within our file structure, for example, we can have our db connection work in one file and api routes in another file, and so on. Such organization is beneficial for both our own and other people's understanding of our code.```
